### PR TITLE
Implement transformation for TensorFlow 2 Map Function (aka tf.map_fn)

### PR DESCRIPTION
**Pull request by rkazants**
**Root cause analysis:** In several TF2 OD models (MobileNet SSD V2, CenterNet) we have a preprocessor block based on tf.map_fn primitive that slices multiple inputs, transform them, and concatenate multiple intermediate outputs at each iteration of While Loop. Currently, the MO tool does not support such block.

**Solution:** Extend existing transformations for TF2 Keras RNN operations that are particular case of tf.map_fn when it has just one sliced input and one concatenated output. The proposed extension detects sub-graphs (from the main-graph and body-graph) for slicing and concatenation and utilizes `axis` attribute of Loop operation for such operations.

**Ticket:** 47280

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests - N/A due to complexity of such unit-tests implementation
* [x]  Framework operation tests
* [x]  Transformation tests - N/A due to complexity of such unit-tests implementation. Now it covers by TF2 Keras RNN layer tests
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list - N/A
* [x]  Guide on how to convert the **public** model - N/A
* [x]  User guide update - N/A

### CODE REVIEW

**Review by achetver — APPROVED** [review]
Looks good to me

File: model-optimizer/extensions/front/tf/MapFNTransformation.py [hunk 1]
```diff
@@ -229,27 +202,46 @@ def transform_keras_rnn_output_concatenation(external_match: dict, internal_matc
...
+                # detects intermediate outputs concatentation by this port and can use Loop axis attribute
+                reserve_node = Loop.get_external_nodes_by_internal_id(loop_node,
+                                                                     internal_match['container'].internal_layer_id)
+                reserve_node = reserve_node[0] if (len(reserve_node) == 1 and
+                                                   reserve_node[0].op == 'TensorListReserve') else None
+                if reserve_node is None:
+                    continue
+                stack_node = Loop.get_external_nodes_by_internal_id(loop_node,
+                                                                   internal_match[
+                                                                       'concatenation_result'].internal_layer_id)
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/MapFNTransformation.py:223]: nit: ugly text wrap.

File: model-optimizer/extensions/ops/loop.py [hunk 2]
```diff
@@ -61,6 +61,27 @@ def get_body_node_by_internal_id(loop_node: Node, internal_id: int):
...
+        from the body graph
+
+        :param loop_node: The Loop node
+        :param internal_layer_id: Internal layer ID of the node in the body graph
+        :return: A list of external nodes (from the main graph) that are connected with a node with
+        internal_layer_id from the body graph
+        """
+        for map_item in loop_node.input_port_map:
+            if map_item['internal_layer_id'] == internal_layer_id \
+                    and not loop_node.in_port(map_item['external_port_id']).disconnected():
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/ops/loop.py:77]: nit: change `not loop_node.in_port(map_item['external_port_id']).disconnected():` to `loop_node.is_in_port_connected(map_item['external_port_id'])`. The proposed solution will not fail if the port with number `map_item['external_port_id']` does not exist in the node. The same comment for line 81.

File: model-optimizer/extensions/front/tf/MapFNTransformation.py [hunk 3]
```diff
@@ -115,51 +94,45 @@ def transform_keras_rnn_input_slicing(external_match: dict, internal_match: dict
...
+            for internal_match in internal_matches:
+                # check if TensorListGetItem from the body graph is connected with TensorListFromTensor
+                # from the main graph. If yes, the transformation detects input slicing by this port
+                # and can use Loop axis attribute
+                unstack_node = Loop.get_external_nodes_by_internal_id(loop_node,
+                                                                     internal_match['tensor_list'].internal_layer_id)
+                unstack_node = unstack_node[0] if (len(unstack_node) == 1
+                                                   and unstack_node[0].op == 'TensorListFromTensor') else None
+                if unstack_node is None:
+                    continue
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/MapFNTransformation.py:112]: nit: Let's log some message before `continue` which include node names which we analyze.

File: model-optimizer/extensions/front/tf/MapFNTransformation.py [hunk 4]
```diff
@@ -229,27 +202,46 @@ def transform_keras_rnn_output_concatenation(external_match: dict, internal_matc
...
+            for internal_match in internal_matches:
+                # check if TensorListReserve from the main graph is connected with Parameter node from the body graph
+                # that is assigned for storing intermediate output results of While Loop. If yes, the transformation
+                # detects intermediate outputs concatentation by this port and can use Loop axis attribute
+                reserve_node = Loop.get_external_nodes_by_internal_id(loop_node,
+                                                                     internal_match['container'].internal_layer_id)
+                reserve_node = reserve_node[0] if (len(reserve_node) == 1 and
+                                                   reserve_node[0].op == 'TensorListReserve') else None
+                if reserve_node is None:
+                    continue
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/MapFNTransformation.py:220]: nit: log some message

File: model-optimizer/extensions/front/tf/MapFNTransformation.py [hunk 5]
```diff
@@ -229,27 +202,46 @@ def transform_keras_rnn_output_concatenation(external_match: dict, internal_matc
...
+                stack_node = Loop.get_external_nodes_by_internal_id(loop_node,
+                                                                   internal_match[
+                                                                       'concatenation_result'].internal_layer_id)
+                stack_node = stack_node if len(stack_node) == 1 else None
+
+                # skip StopGradient node if it exists between While loop output port and TensorListStack operation
+                if stack_node is None:
+                    continue
+                if stack_node[0].op == 'StopGradient':
+                    stack_node = [dest.node for dest in stack_node[0].out_port(0).get_destinations()]
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/front/tf/MapFNTransformation.py:230]: I would recommend to skip all nodes which are marked with `identity=True` attribute in the while loop.  Use method `skip_nodes_by_condition` from the ObjectDetectionAPI.py file. You can move it to some common graph functions module.

**Review by lazarevevgeny — COMMENTED** [review]
There are a number of minor comments, but in general looks good. However, I would really want to see a new layer test covering the added functionality:

1. A test case where a model has one input which is consumed with the TF `While` operation which produces 2 outputs.
2. A test case where a model has two inputs which are consumed with the TF `While` operation which produces 2 outputs.

If one of these test cases is already covered with the existing Keras RNN-like tests that is fine. But the case which was not previously supported should be added to the **pre-commit** layer tests.

File: model-optimizer/extensions/front/tf/MapFNTransformation.py [hunk 6]
```diff
@@ -115,51 +94,45 @@ def transform_keras_rnn_input_slicing(external_match: dict, internal_match: dict
...
+            body_graph = loop_node['body']
+            body_pattern = MapFNInputSlicing.get_body_pattern()
+            internal_matches = find_subgraph_match_to_pattern(body_graph, body_pattern)
+
+            for internal_match in internal_matches:
+                # check if TensorListGetItem from the body graph is connected with TensorListFromTensor
+                # from the main graph. If yes, the transformation detects input slicing by this port
+                # and can use Loop axis attribute
+                unstack_node = Loop.get_external_nodes_by_internal_id(loop_node,
+                                                                     internal_match['tensor_list'].internal_layer_id)
```
[the remark this answers is not in the document]
Comment (rkazants) [reply model-optimizer/extensions/front/tf/MapFNTransformation.py]: Fixed. Thanks.

File: model-optimizer/extensions/front/tf/MapFNTransformation.py [hunk 7]
```diff
@@ -229,27 +202,46 @@ def transform_keras_rnn_output_concatenation(external_match: dict, internal_matc
...
+            body_graph = loop_node['body']
+            body_pattern = MapFNOutputConcatenation.get_body_pattern()
+            internal_matches = find_subgraph_match_to_pattern(body_graph, body_pattern)
+
+            for internal_match in internal_matches:
+                # check if TensorListReserve from the main graph is connected with Parameter node from the body graph
+                # that is assigned for storing intermediate output results of While Loop. If yes, the transformation
+                # detects intermediate outputs concatentation by this port and can use Loop axis attribute
+                reserve_node = Loop.get_external_nodes_by_internal_id(loop_node,
+                                                                     internal_match['container'].internal_layer_id)
```
[the remark this answers is not in the document]
Comment (rkazants) [reply model-optimizer/extensions/front/tf/MapFNTransformation.py]: Fixed. Thanks.

File: model-optimizer/extensions/front/tf/MapFNTransformation.py (hunk 1 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/MapFNTransformation.py]: Fixed. Thanks.

File: model-optimizer/extensions/ops/loop.py (hunk 2 above)
Comment (rkazants) [reply model-optimizer/extensions/ops/loop.py]: Fixed. Thanks.

File: model-optimizer/extensions/front/tf/MapFNTransformation.py (hunk 3 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/MapFNTransformation.py:118]: Fixed. Thanks.

File: model-optimizer/extensions/front/tf/MapFNTransformation.py (hunk 4 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/MapFNTransformation.py:229]: Fixed. Thanks.

File: model-optimizer/extensions/front/tf/MapFNTransformation.py (hunk 5 above)
Comment (rkazants) [reply model-optimizer/extensions/front/tf/MapFNTransformation.py]: Done. Thanks.

**Comment by rkazants** [comment]
> There are a number of minor comments, but in general looks good. However, I would really want to see a new layer test covering the added functionality:
> 
> 1. A test case where a model has one input which is consumed with the TF `While` operation which produces 2 outputs.
> 2. A test case where a model has two inputs which are consumed with the TF `While` operation which produces 2 outputs.
> 
> If one of these test cases is already covered with the existing Keras RNN-like tests that is fine. But the case which was not previously supported should be added to the **pre-commit** layer tests.

Thanks for your point. I will add it.

**Review by lazarevevgeny — APPROVED** [review]

**Review by iimironov — APPROVED** [review]

**Review by pavel-esir — APPROVED** [review]
